_posts/en/2020-01-09-rxjs-introduction.md

```diff
@@ -0,0 +1,248 @@
+
+### subscribe() and unsubscribe()
+As a parameter, the subscribe function expects an object that defines three functions: _next_, _error_ and _complete_.
+```typescript
+const subscription = numbers$.subscribe({
+    next(value) {},
+    error(err) {},
+    complete() {}
+});
+```
+
+Each of these functions is a callback that is called at given moments of data flowing through the stream. The _next(value)_ function is called every time the stream emits a single value, which in this case means that the _next()_ function will be called 5 times, 1 time for every digit from 1 to 5. The _error(err)_ callback will be called when the stream is unnaturally closed or disrupted. Complete is the last callback that is called once the stream has been closed.
+
+Calling the _subscribe()_ function adds you to the list of observers of a given stream.
+
+When subscribing as an observer to a given observable, you are given a subscription-type  object on which you can call the _unsubscribe()_ method to remove yourself from the list of observers.
+
+Unsubscribing is crucial in the case of hot streams, as they are mostly infinite, meaning they emit the value for a potentially infinite amount of time. If you forget to unsubscribe from the list of observers of a such a stream, the reference to the observer you’ve created will exist throughout the whole lifecycle of the application, creating an infinite hole in the memory, which in extreme cases might lead to the death of the tab that the application is running in.
+
+## Subject - creating your own streams
+RxJS enables you to create your own streams and to do so you can use a subject-type object. Such a stream is potentially infinite, emitting new values at what you consider important points of the application. A subject is created in the same way as other objects:
+```typescript
+const subject$ = new Subject<number>();
+```
+Now you can subscribe to the stream as an observer:
+```typescript
+subject$.asObservable()
+    .subscribe((value) => console.log('value from subject$: ', value))
+```
+The _subject$_ variable now provides you with the _next(value: number)_ method, which allows you to send out a new numeric value to all observers, like this:
+```typescript
+subject$.next(5);
+// value from subject$: 5
+```
+
+The stream you’ve just created is infinite, so you need to remember to unsubscribe once you finish listening to it. You can also close the stream with the _complete()_ command.
+
+In this example, a numerical value ‘5’ was emitted and received by a single observer. If no observers exist at the time of emitting a new value, it is lost. However, if this value is important to you and you don’t want to lose, you can use specific subject-type extending objects.
```

Comment:
w sumie to nie wiem czy to ważne, ale zauważyłem, że w artykule Tomka preferowane są double quotes a w tym single

Reply:
To w sumie pierdoła, ale przejechałem przez te wpisy po polsku i użyłem taich cudzysłowów, jakie są tam.

Reply:
Tak jak pisałem, nie wiem czy to ważne, czy nie - po prostu rzuciło mi się w oczy :)